File: app/src/main/java/me/instabattle/app/adapters/UserEntryListAdapter.java
package me.instabattle.app.adapters;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;

import me.instabattle.app.managers.BitmapCallback;
import me.instabattle.app.models.Battle;
import me.instabattle.app.models.Entry;
import me.instabattle.app.R;
import me.instabattle.app.settings.State;
import me.instabattle.app.activities.BattleActivity;
import me.instabattle.app.activities.MapActivity;
import me.instabattle.app.activities.MyProfileActivity;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserEntryListAdapter extends BaseAdapter {

    private static final String TAG = "UserEntryListManager";

    private List<Entry> entries;
    private Context context;
    private LayoutInflater inflater;

    public UserEntryListAdapter(Context context, List<Entry> entries) {
        this.context = context;
        this.entries = entries;
        this.inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    }

    @Override
    public int getCount() {
        return entries.size();
    }

    @Override
    public Object getItem(int position) {
        return entries.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        final View res = convertView != null ? convertView :
                inflater.inflate(R.layout.user_entry_list_item, parent, false);

        Entry entry = entries.get(position);

        entry.getPhotoAndDo(new BitmapCallback() {
            @Override
            public void onResponse(final Bitmap photo) {
                ((Activity)context).runOnUiThread(() -> ((ImageView) res.findViewById(R.id.userEntryListItemImage)).setImageBitmap(photo));
            }

            @Override
            public void onFailure(Exception e) {
                //TODO
                Log.e(TAG, "Can't get entry photo");
            }
        });

        ((TextView) res.findViewById(R.id.userEntryListItemDate)).setText(
                (new SimpleDateFormat("dd/MM/yyyy", Locale.US)).format(entry.getCreatedOn()));
        ((TextView) res.findViewById(R.id.userEntryListItemUpvotes)).setText(entry.getRating() + " points");

        entry.getBattleAndDo(new Callback<Battle>() {
            @Override
            public void onResponse(Call<Battle> call, Response<Battle> response) {
                final Battle battle = response.body();

                ((TextView) res.findViewById(R.id.userEntryListItemTitle)).setText(battle.getName());

                res.findViewById(R.id.userEntryListItemViewBtn).setOnClickListener(v -> {
                    State.chosenBattle = battle;
                    BattleActivity.Companion.setGotHereFrom(MyProfileActivity.class);
                    Intent viewBattle = new Intent(context, BattleActivity.class);
                    context.startActivity(viewBattle);
                });

                res.findViewById(R.id.userEntryListItemViewOnMapBtn).setOnClickListener(v -> {
                    MapActivity.viewPoint = battle.getLocation();
                    MapActivity.viewZoom = MapActivity.DEFAULT_ZOOM;
                    MapActivity.gotHereFrom = MyProfileActivity.class;
                    Intent viewMap = new Intent(context, MapActivity.class);
                    context.startActivity(viewMap);
                });
            }

            @Override
            public void onFailure(Call<Battle> call, Throwable t) {
                Log.e(TAG, "cant get entries battle: " + t);
                //TODO
            }
        });
        return res;
    }
}

File: app/src/main/kotlin/me/instabattle/app/adapters/UserEntryListAdapter.kt
package me.instabattle.app.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import me.instabattle.app.R
import me.instabattle.app.activities.BattleActivity
import me.instabattle.app.activities.MapActivity
import me.instabattle.app.activities.MyProfileActivity
import me.instabattle.app.managers.PhotoManager
import me.instabattle.app.models.Battle
import me.instabattle.app.models.Entry
import me.instabattle.app.settings.State
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.error
import org.jetbrains.anko.startActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.text.SimpleDateFormat
import java.util.*

class UserEntryListAdapter(private val ctx: Context, private val entries: List<Entry>) : BaseAdapter(), AnkoLogger {
    private val inflater = ctx.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    override fun getCount(): Int {
        return entries.size
    }

    override fun getItem(position: Int): Any {
        return entries[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val res = convertView ?: inflater.inflate(R.layout.user_entry_list_item, parent, false)

        val entry = entries[position]

        PhotoManager.getPhotoInto(ctx, entry.imageName!!, res.findViewById<ImageView>(R.id.listEntryImage))

        res.findViewById<TextView>(R.id.listEntryUpvotes).text = SimpleDateFormat("dd/MM/yyyy", Locale.US).format(entry.createdOn)
        res.findViewById<TextView>(R.id.userEntryListItemUpvotes).text = entry.rating.toString() + " points"

        entry.getBattleAndDo(object : Callback<Battle> {
            override fun onResponse(call: Call<Battle>, response: Response<Battle>) {
                val battle = response.body()

                res.findViewById<TextView>(R.id.userEntryListItemTitle).text = battle!!.name

                res.findViewById<View>(R.id.userEntryListItemViewBtn).setOnClickListener { v ->
                    State.chosenBattle = battle
                    BattleActivity.gotHereFrom = MyProfileActivity::class.java
                    ctx.startActivity<BattleActivity>()
                }

                res.findViewById<View>(R.id.userEntryListItemViewOnMapBtn).setOnClickListener { v ->
                    MapActivity.viewPoint = battle.location
                    MapActivity.viewZoom = MapActivity.DEFAULT_ZOOM
                    ctx.startActivity<MapActivity>()
                }
            }

            override fun onFailure(call: Call<Battle>, t: Throwable) {
                error("can't get entries battle: $t")
            }
        })
        return res
    }
}
